flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java

```diff
@@ -225,6 +226,21 @@ public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
 		}
 	}
 
+	/**
+	 * Starting from the second operator, go forward through the operator chain to notify
+	 * each operator that its input has ended.
+	 *
+	 * @throws Exception if some exception happens in the endInput function of an operator.
+	 */
+	public void endOperatorInputs() throws Exception {
+		for (int i = allOperators.length - 2; i >= 0; i--) {
```

Comment:
The `endInput()` of the head operator has been called in `xInputProcessor`. It's called in `xInputProcessor` because the `endInput()` of the two input operator must be called immediately when one of both inputs arrives the end.

Reply:
I'll change the method name `endOperatorInputs` to the appropriate one (for example, `endNonHeadOperatorInputs`).

Reply:
> I think you could simplify the code if you iterated through all operators in OperatorChain (not starting from the second one), check each one of them if there are instanceof BoundedOneInput and end them if they are.
>
>This would work for:
>

> - OneInputStreamTask - because all of them are/can be BoundedOneInput
> - SourceStreamTask - because head will not be BoundedOneInput
> - TwoInputStreamTask - because head will not be BoundedOneInput

You are right. However, if chaining non-header two input operator is allowed in the future, it will have problems.

Reply:
Maybe we can use the following structure?